Client/Frontend/Home/ActivityStreamPanel.swift

```diff
@@ -0,0 +1,337 @@
+        }
+    }
+
+    private func refreshTopSites(frecencyLimit: Int) {
+        dispatch_async(dispatch_get_main_queue()) {
+            // Reload right away with whatever is in the cache, then check to see if the cache is invalid.
+            // If it's invalid, invalidate the cache and requery. This allows us to always show results
+            // immediately while also loading up-to-date results asynchronously if needed.
+            self.reloadTopSitesWithLimit(frecencyLimit) >>> {
+                self.profile.history.updateTopSitesCacheIfInvalidated() >>== { dirty in
+                    if dirty {
+                        self.reloadTopSitesWithLimit(frecencyLimit)
+                    }
+                }
+            }
+        }
+    }
+
+    private func reloadTopSitesWithLimit(limit: Int) -> Success {
+        return self.profile.history.getTopSitesWithLimit(limit).bindQueue(dispatch_get_main_queue()) { result in
+            if let data = result.successValue {
+                self.topSites = data.asArray().map { site in
+                    if let favURL = site.icon?.url {
+                        return TopSiteItem(urlTitle: site.tileURL.extractDomainName(), faviconURL: NSURL(string: favURL)!, siteURL: site.tileURL)
+                    }
+                    return TopSiteItem(urlTitle: site.tileURL.extractDomainName(), faviconURL: nil, siteURL: site.tileURL)
+                }
+                self.topSiteHandler.content = self.topSites
+                self.topSiteHandler.urlPressedHandler = { [unowned self] url in
+                    self.showSiteWithURL(url)
+                }
+                self.topSiteHandler.currentTraits = self.traitCollection
+                self.tableView.reloadData()
+            }
+            return succeed()
+        }
+    }
+
+    private func reloadRecentHistoryWithLimit(limit: Int) -> Success {
+        return self.profile.history.getSitesByLastVisit(limit).bindQueue(dispatch_get_main_queue()) { result in
+            if let data = result.successValue {
+                self.history = data.asArray()
+                self.tableView.reloadData()
+            }
+            return succeed()
+        }
+    }
+}
+
+// MARK: - HomePanel Protocol
+extension ActivityStreamPanel: HomePanel {
+
+    func endEditing() {
+
+    }
+
+}
+
+// MARK: - Section Header View
+struct ASHeaderViewUX {
+    static let SeperatorColor =  UIColor(rgb: 0xedecea)
+    static let TextFont = DynamicFontHelper.defaultHelper.DefaultSmallFontBold
+    static let SeperatorHeight = 1
+    static let Insets: CGFloat = 20
+    static let TitleTopInset: CGFloat = 5
+}
+
+class ASHeaderView: UIView {
+    lazy private var titleLabel: UILabel = {
+        let titleLabel = UILabel()
+        titleLabel.text = self.title
+        titleLabel.textColor = UIColor.grayColor()
+        titleLabel.font = ASHeaderViewUX.TextFont
+        return titleLabel
+    }()
+
+    var title: String = "" {
+        willSet(newTitle) {
```

Comment:
Nit: Not sure if it makes a difference but logically I think of this as 'after setting the title, also set the label'  which would be using `didSet` instead `willSet`.